File: src/transcript.mjs
/**
 * Tool-call counter.
 *
 * Session transcripts grow to megabytes, and the statusline renders on every
 * keystroke-ish event, so re-reading the whole file is not an option. We keep a
 * per-session cursor (byte offset + running count) and only ever read the bytes
 * appended since the last render.
 */

import { closeSync, existsSync, openSync, readSync, readFileSync, renameSync, statSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { cacheDir, ensureCacheDir } from "./config.mjs";

const TOOL_USE_MARKER = /"type"\s*:\s*"tool_use"/g;
/** Longest marker we could split across a read boundary; overlap by this much. */
const MARKER_SLACK = 32;

function cursorPath(sessionId, env = process.env) {
  const safe = String(sessionId).replace(/[^A-Za-z0-9._-]/g, "_");
  return join(cacheDir(env), `tools-${safe}.json`);
}

function readCursor(path) {
  try {
    const parsed = JSON.parse(readFileSync(path, "utf8"));
    if (typeof parsed?.offset === "number" && typeof parsed?.count === "number") return parsed;
  } catch {
    /* a missing or corrupt cursor just means we start from zero */
  }
  return null;
}

function writeCursor(path, cursor, env = process.env) {
  try {
    ensureCacheDir(env);
    const tmp = `${path}.${process.pid}.tmp`;
    writeFileSync(tmp, JSON.stringify(cursor));
    renameSync(tmp, path);
  } catch {
    /* without a cursor we simply recount next time */
  }
}

function countMarkers(text) {
  TOOL_USE_MARKER.lastIndex = 0;
  let count = 0;
  while (TOOL_USE_MARKER.exec(text) !== null) count += 1;
  return count;
}

/**
 * @returns {number|null} tool calls in this session, or null when the
 *   transcript is unavailable.
 */
export function countToolCalls(transcriptPath, sessionId, env = process.env) {
  if (!transcriptPath || !existsSync(transcriptPath)) return null;

  let size;
  try {
    size = statSync(transcriptPath).size;
  } catch {
    return null;
  }

  const path = cursorPath(sessionId ?? transcriptPath, env);
  const cached = readCursor(path);
  // A shrunken file means the transcript was rotated or replaced — recount.
  const resume = cached && cached.size <= size && cached.offset <= size ? cached : null;
  let offset = resume?.offset ?? 0;
  let count = resume?.count ?? 0;

  if (offset >= size) return count;

  let fd;
  try {
    fd = openSync(transcriptPath, "r");
    // Re-read a few bytes before the cursor so a marker straddling the previous
    // boundary is not missed; the overlap is trimmed at the first newline.
    // All arithmetic below stays in BYTES — transcripts contain multi-byte
    // UTF-8, so string indices would drift from file offsets.
    const start = Math.max(0, offset - MARKER_SLACK);
    const buffer = Buffer.allocUnsafe(size - start);
    const read = readSync(fd, buffer, 0, buffer.length, start);
    const view = buffer.subarray(0, read);

    let dataStart = offset - start;
    if (start < offset) {
      const firstNewline = view.indexOf(0x0a);
      if (firstNewline === -1) return count;
      dataStart = firstNewline + 1;
    }

    // Only count whole lines; a half-written final line is left for next time.
    const lastNewline = view.lastIndexOf(0x0a);
    if (lastNewline === -1 || lastNewline < dataStart) return count;
    count += countMarkers(view.toString("utf8", dataStart, lastNewline + 1));
    offset = start + lastNewline + 1;
  } catch {
    return count || null;
  } finally {
    if (fd !== undefined) {
      try {
        closeSync(fd);
      } catch {
        /* ignore */
      }
    }
  }

  writeCursor(path, { offset, count, size }, env);
  return count;
}

File: transcript.go
package main

import (
	"bytes"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

var toolUseMarker = regexp.MustCompile(`"type"\s*:\s*"tool_use"`)

type transcriptCursor struct {
	Offset int64 `json:"offset"`
	Count  int   `json:"count"`
	Size   int64 `json:"size"`
}

// countToolCalls counts `"type": "tool_use"` markers in the session transcript
// (a JSONL file that grows to megabytes) by reading only the bytes appended
// since the last render, with a per-session cursor in the cache dir. A cursor
// that does not fit the current file (size shrank, offset negative or past the
// end) is dropped, which is how a rotated transcript restarts from zero. Only
// whole lines count and the cursor is left just past the last newline read, so
// the next frame resumes on a line boundary and the trailing partial line
// waits. All arithmetic stays in bytes: transcripts contain multi-byte UTF-8.
func countToolCalls(transcriptPath, sessionID string, env Env) (int, bool) {
	if transcriptPath == "" {
		return 0, false
	}
	info, err := os.Stat(transcriptPath)
	if err != nil {
		return 0, false
	}
	size := info.Size()
	if sessionID == "" {
		sessionID = transcriptPath
	}
	var safe strings.Builder
	for _, r := range sessionID {
		if r < 128 && ((r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '.' || r == '_' || r == '-') {
			safe.WriteRune(r)
		} else {
			safe.WriteByte('_')
		}
	}
	cursorPath := filepath.Join(env.cacheDir(), "tools-"+safe.String()+".json")

	var offset int64
	var count int
	var cached struct {
		Offset *int64 `json:"offset"`
		Count  *int   `json:"count"`
		Size   *int64 `json:"size"`
	}
	if data, err := os.ReadFile(cursorPath); err == nil && json.Unmarshal(data, &cached) == nil &&
		cached.Offset != nil && cached.Count != nil && cached.Size != nil &&
		*cached.Size <= size && *cached.Offset >= 0 && *cached.Offset <= size {
		offset = *cached.Offset
		count = *cached.Count
	}
	if offset >= size {
		return count, true
	}

	file, err := os.Open(transcriptPath)
	if err != nil {
		if count > 0 {
			return count, true
		}
		return 0, false
	}
	defer file.Close()
	buffer := make([]byte, size-offset)
	read, readErr := file.ReadAt(buffer, offset)
	if readErr != nil && readErr != io.EOF {
		if count > 0 {
			return count, true
		}
		return 0, false
	}
	view := buffer[:read]
	lastNewline := bytes.LastIndexByte(view, '\n')
	if lastNewline == -1 {
		return count, true
	}
	count += len(toolUseMarker.FindAll(view[:lastNewline+1], -1))
	offset += int64(lastNewline) + 1

	cursor := transcriptCursor{Offset: offset, Count: count, Size: size}
	if data, marshalErr := json.Marshal(cursor); marshalErr == nil {
		env.ensureCacheDir()
		tmp := cursorPath + "." + strconv.Itoa(os.Getpid()) + ".tmp"
		if writeErr := os.WriteFile(tmp, data, 0o666); writeErr == nil {
			_ = os.Rename(tmp, cursorPath)
		}
	}
	return count, true
}
